base/regex.jl

```diff
@@ -152,8 +152,10 @@ function match(re::Regex, str::UTF8String, idx::Integer, add_opts::UInt32=UInt32
     RegexMatch(mat, cap, ovec[1]+1, off, re)
 end
 
+_utf8(str) = utf8(str)
```

Comment:
This doesn't seem right, because `_utf8(::SubString{UTF8String})` still makes a copy.  You could use
```jl
_utf8(str) = utf8(str)
_utf8{T<:ByteString}(str::SubString{T}) = convert(SubString{UTF8String}, str)
```
where the `convert` is a no-op for `str::SubString{UTF8String}`.

And, for constency, shouldn't it be `match{T<:ByteString}(re::Regex, str::Union{T,SubString{T}}, ...)` so that `SubString` of e.g. `UTF16String` will give a `MethodError` here?   It seems weird to allow either a `ByteString` or an arbitrary `SubString` of any string type.